flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java

```diff
@@ -90,6 +90,25 @@ public boolean snapshotExists(long snapshotId) {
         }
     }
 
+    public @Nullable Long latestCompactedSnapshotId() {
+        try {
+            Iterator<Snapshot> iterator = snapshots();
+            Long maxCompactedSnapshotId = null;
+            while (iterator.hasNext()) {
+                Snapshot snapshot = iterator.next();
+                if (snapshot.commitKind() == Snapshot.CommitKind.COMPACT) {
+                    if (maxCompactedSnapshotId == null || snapshot.id() > maxCompactedSnapshotId) {
+                        maxCompactedSnapshotId = snapshot.id();
+                    }
+                }
+            }
+
+            return maxCompactedSnapshotId;
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to find latest compacted snapshot id", e);
+        }
```

Comment:
This implementation is very inefficient with thousands of snapshots. Try iterating backwards from the latest snapshot. When you spot a COMPACT snapshot you can return directly.